File: sample/src/main/java/cc/coopersoft/android/keycloak/MainActivity.java
package cc.coopersoft.android.keycloak;

import android.content.Context;
import android.content.Intent;
import android.os.CountDownTimer;
import android.os.Looper;
import android.text.Editable;
import android.text.TextWatcher;
import android.view.Gravity;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import android.os.Bundle;
import androidx.databinding.DataBindingUtil;
import cc.coopersoft.accesstoken.api.callback.CodeRequestCallback;
import cc.coopersoft.accesstoken.api.callback.ErrorResult;
import cc.coopersoft.accesstoken.api.callback.VerificationCallback;
import cc.coopersoft.android.keycloak.databinding.ActivityMainBinding;

public class MainActivity extends AppCompatActivity {

  private ActivityMainBinding binding;

  @Override
  protected void onCreate(Bundle savedInstanceState) {
    super.onCreate(savedInstanceState);
    binding = DataBindingUtil.setContentView(this,R.layout.activity_main);
    initView();
  }

  private void toast(String text) {
    Toast toast = Toast.makeText(getApplicationContext(), text, Toast.LENGTH_LONG);
    toast.setGravity(Gravity.TOP, 0, 0);
    toast.show();
  }

  private void initView() {

    binding.phoneNumber.addTextChangedListener(new TextWatcher() {
      @Override
      public void beforeTextChanged(CharSequence charSequence, int i, int i1, int i2) {

      }

      @Override
      public void onTextChanged(CharSequence charSequence, int i, int i1, int i2) {
        binding.requestAuthCode.setEnabled(charSequence.toString().length() == 11);
      }

      @Override
      public void afterTextChanged(Editable editable) {

      }
    });

    // send  verification code
    binding.requestAuthCode.setOnClickListener(view -> {
      binding.authCode.setFocusable(true);
      binding.requestAuthCode.setEnabled(false);
      KeycloakClient.accessTokenHolder().sendVerificationCode(binding.phoneNumber.getText().toString(), new CodeRequestCallback() {
        @Override
        public void onFailure(String phoneNumber, ErrorResult error) {
          binding.requestAuthCode.setEnabled(true);
          runOnUiThread(()-> toast("send code fail!"));
        }

        @Override
        public void onSuccess(String phoneNumber, Result result) {
          Looper.prepare();
          new CountDownTimer(result.getExpiresIn() * 1000, 1000) {
            @Override
            public void onTick(long millisUntilFinished) {
              binding.requestAuthCode.setEnabled(false);
              binding.requestAuthCode.setText("send(" + millisUntilFinished / 1000 + ")");

            }

            @Override
            public void onFinish() {
              binding.requestAuthCode.setEnabled(true);
              binding.requestAuthCode.setText("resend");

            }
          }.start();
          Looper.loop();
        }
      });
    });

    // verification action
    binding.verification.setOnClickListener(view -> {
      Runnable verificationTask = () -> KeycloakClient.accessTokenHolder().verificationCode(binding.phoneNumber.getText().toString(), binding.authCode.getText().toString(), new VerificationCallback() {
        @Override
        public void onError(String phoneNumber, ErrorResult error) {
          runOnUiThread(()-> toast("valid fail!" + error.getErrorDescription()));
        }

        @Override
        public void onFailure(String phoneNumber) {
          runOnUiThread(()-> toast("valid fail!"));
        }

        @Override
        public void onSuccess(String phoneNumber) {
          runOnUiThread(()-> toast("valid success"));
        }
      });

      new Thread(verificationTask).start();


    });


    //logout action
    binding.logout.setOnClickListener(view -> {
      KeycloakClient.accessTokenHolder().clearToken();
      LoginActivity.start(this);
    });

  }

  public static void start(Context context) {
    context.startActivity(new Intent(context, MainActivity.class));
  }


}

File: sample/src/main/java/cc/coopersoft/android/keycloak/MainActivity.kt
package cc.coopersoft.android.keycloak

import android.content.Context
import android.content.Intent
import android.os.Bundle
import android.os.CountDownTimer
import android.os.Looper
import android.text.Editable
import android.text.TextWatcher
import android.util.Log
import android.view.Gravity
import android.view.View
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import cc.coopersoft.accesstoken.api.callback.CodeRequestCallback
import cc.coopersoft.accesstoken.api.callback.ErrorResult
import cc.coopersoft.accesstoken.api.callback.VerificationCallback
import cc.coopersoft.android.keycloak.KeycloakClient.accessTokenHolder
import cc.coopersoft.android.keycloak.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private var binding: ActivityMainBinding? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        binding = DataBindingUtil.setContentView(this, R.layout.activity_main)
        initView()
    }

    private fun toast(text: String) {
        val toast = Toast.makeText(applicationContext, text, Toast.LENGTH_LONG)
        toast.setGravity(Gravity.TOP, 0, 0)
        toast.show()
    }

    private fun initView() {
        binding?.phoneNumber?.addTextChangedListener(object : TextWatcher {
            override fun beforeTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {}
            override fun onTextChanged(charSequence: CharSequence, i: Int, i1: Int, i2: Int) {
                binding?.requestAuthCode?.isEnabled = charSequence.toString().length == 11
            }

            override fun afterTextChanged(editable: Editable) {}
        })

        // send  verification code
        binding?.requestAuthCode?.setOnClickListener { view: View? ->
            binding?.authCode?.isFocusable = true
            binding?.requestAuthCode?.isEnabled = false
            accessTokenHolder()!!
                .sendVerificationCode(
                    binding?.phoneNumber?.text.toString(),
                    object : CodeRequestCallback {
                        override fun onFailure(phoneNumber: String?, error: ErrorResult) {
                            binding?.requestAuthCode?.isEnabled = true
                            error.error?.let { Log.e(TAG, it) }
                            runOnUiThread { toast("send code fail!") }
                        }

                        override fun onSuccess(
                            phoneNumber: String?,
                            result: CodeRequestCallback.Result
                        ) {
                            Looper.prepare()
                            object : CountDownTimer(result!!.expiresIn * 1000, 1000) {
                                override fun onTick(millisUntilFinished: Long) {
                                    binding?.requestAuthCode?.isEnabled = false
                                    binding?.requestAuthCode?.text = "send(" + millisUntilFinished / 1000 + ")"
                                }

                                override fun onFinish() {
                                    binding?.requestAuthCode?.isEnabled = true
                                    binding?.requestAuthCode?.text = "resend"
                                }
                            }.start()
                            Looper.loop()
                        }
                    })
        }

        // verification action
        binding?.verification?.setOnClickListener { view: View? ->
            val verificationTask = Runnable {
                accessTokenHolder()!!
                    .verificationCode(
                        binding?.phoneNumber?.text.toString(),
                        binding?.authCode?.text.toString(),
                        object : VerificationCallback {
                            override fun onError(phoneNumber: String?, error: ErrorResult) {
                                runOnUiThread { toast("valid fail!" + error.errorDescription) }
                            }

                            override  fun onFailure(phoneNumber: String?) {
                                runOnUiThread { toast("valid fail!") }
                            }

                            override   fun onSuccess(phoneNumber: String?) {
                                runOnUiThread { toast("valid success") }
                            }
                        })
            }
            Thread(verificationTask).start()
        }


        //logout action
        binding?.logout?.setOnClickListener { view: View? ->
            accessTokenHolder()!!.clearToken()
            LoginActivity.start(this)
        }
    }

    companion object {
        const val TAG = "MainActivity"
        @JvmStatic
        fun start(context: Context) {
            context.startActivity(Intent(context, MainActivity::class.java))
        }
    }
}